### ISSUE TITLE
Feature request: read raw SF2 data from memory

### ISSUE THREAD

**Issue by sykhro** [FluidSynth/fluidsynth#241]
Hello,
while cleaning up a Linux port of a certain tool, I hit a particular use case: I have to load SF2 data from memory. I can read MIDI from a buffer just fine, but when it comes to loading a soundfont there's no direct way. The problem was somewhat fixed by implementing [this](https://github.com/vgmtrans/vgmtrans/tree/master/lib/fluidsynth/mem_sfloader/mem_sfloader.c), but it doesn't feel like a great solution because the code itself is very old and it's more of a hack around the api. 
I understand that this will probably be turned down as it won't probably be used by a lot of people, but I'm not sure if it's worth re-implementing an entire loader for a relatively minor addition

**Comment by derselbst** [comment]
Implementing another sfont loader wont happen. However one might think about creating a user supplied struct of function pointers, that replicate the behaviour of fopen(), fread(), etc. Something like this: 

https://gitlab.kode54.net/kode54/psflib/blob/7da887b3a41ece72b2461cbf8876af6dda2738fc/psflib.h#L36-55

So instead that fluidsynth calls `fread()` directly, it would call the user supplied function pointer.

There are currently other open issues around here, so I wont have time implementing that. Feel free to draft a PR though.

**Comment by sykhro** [comment]
That seems like a good solution; I'll look into it whenever I can

**Comment by sykhro** [comment]
finally had time to try to do something about this, but there are a few problems. first off, I'm unsure to what the callbacks struct should be passed to (I was thinking of adding it to the loader struct, and then referencing it when needed. I wanted to avoid changing `new_fluid_defsfloader`'s prototype though. there's a way around it with variatic functions, but it feel overcomplicated). secondly, the filename remains a problem, because it's used to set the sfont filename. this ends up in the sfont filename always being "RIFFE", because it reads from the file header, and it stops at null bytes.

**Comment by derselbst** [comment]
You could add the callback struct as a member to `struct _fluid_sfloader_t`. `new_fluid_defsfloader()` would create the `_fluid_sfloader_t` and init the callbacks to the standard C file api functions. After exposing `new_fluid_defsfloader()` to public API, the user could retrieve this sturct, manipulate the callbacks and then add this loader via `fluid_synth_add_sfloader()`. Internally you will only have to pass the callback struct to `fluid_defsfont_load()` and every other function that would need to call fopen(), fread() and friends.

When the user then wants to load a sf2 from memory he could:

```
char abused_filename[64];
sprintf(abused_filename, "&%p", pointer_to_sf2_in_mem);
fluid_synth_sfload(synth, abused_filename, FALSE);
```

Ultimately `abused_filename` gets interpreted by any user supplied function fopen() function, in which you would probably want to use `scanf()` to get your pointer back again ;)

Quick thought... hope that makes sense.

> this ends up in the sfont filename always being "RIFFE", because it reads from the file header, and it stops at null bytes.

Dont quite understand what you mean here, sry.

**Comment by sykhro** [comment]
Yep, that's more or less what I had in mind. I was afraid of compatibility problems but realised just now that most stuff in fluid_defsfont is not public. Sorry, I'm not too familiar with the lib structure yet.

> Dont quite understand what you mean here, sry.
Whoops, nevermind that. I got some code mixed up at some point

**Comment by mawe42** [comment]
Isn't @sykhro use-case exactly what the fluid_ramsfont loader was written for? Or is that code just too old to be useful?

**Comment by derselbst** [comment]
@mawe42 I'm understanding @sykhro's use case as that he has a whole SF2 file in memory and wants to load it from there (same as it can be done with MIDI files).

Regarding ramsfont I cant tell. All that I see in ramsfont is a huge code duplication from defsfont with some functional limitations, that is hard to maintain. I was even thinking about removing it (after having this discussed on fluid-dev ofc). Element once told me about it:

> I remember a contributor added this (not part of the core team).  I also thought it wasn't so useful, especially in the case of Swami [...], which has it's own instrument processing using the FluidSynth SoundFont loader API.  I believe the ramsfont thing is for dynamically constructing SoundFont data, without having to load it from a file (as libInstPatch/Swami does).  I've never used ramsfont myself.

So if @sykhro could make use of ramsfont that would be great. It'd give me an idea of its original purpose and a reason to further maintain it. ...The question would still be if one hadnt been able to achieve the same functionality by exposing defsfont though.

**Comment by sykhro** [comment]
I had the feeling ramsfont wasn't really a memory loader (docs didn't help much); before opening this issue I tried making use of it, but it didn't seem to fit my use case that much. In my scenario, SF2 are loaded in memory and are ready to be used, which is why upstream mantainers [did this](https://github.com/vgmtrans/vgmtrans/blob/master/lib/fluidsynth/mem_sfloader/mem_sfloader.c). Overriding the stdio functions is the most straightforward solution imo, I hope I can make a pr asap.

**Comment by mikelow** [comment]
I wrote the hacky solution you referenced. At the time this repo didn't exist, so there wasn't an easy means to put up a PR / communicate with the devs. IIRC, my solution isn't doing anything special beyond duplicating the default sfont loader and wrapping the fread and fseek calls into new functions (and treating the filename pointer as a data pointer, which would go into the fopen hook), so derselbst's proposal sounds quite viable.

**Comment by derselbst** [comment]
Feature implemented in #291 and will be part of the next major release. Will leave this ticket open for a couple of days (weeks) for testing.

**Comment by derselbst** [comment]
Due to the refactoring of the soundfont loader API in #334, some changes were introduced also affecting this issue. Instead of using some public struct to specify the callbacks, now there is a function `fluid_sfloader_set_callbacks()`. The [latest API doc is available here](http://www.fluidsynth.org/api-unstable/sfont_8h.html), the [latest usage example here](http://www.fluidsynth.org/api-unstable/fluidsynth_sfload_mem_8c-example.html).

I consider this feature to be implemented now, therefore I'm closing this. As usual in case of a bug, just file a new issue, any general questions please discuss at the mailing list.
